pubsublite/types.go

```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+
+package pubsublite
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+// LocationPath stores a path consisting of a project and zone.
+type LocationPath struct {
+	// A Google Cloud project. The project ID (e.g. "my-project") or the project
+	// number (e.g. 987654321) can be provided.
+	Project string
+
+	// A Google Cloud zone, for example "us-central1-a".
+	// See https://cloud.google.com/pubsub/lite/docs/locations for the list of
+	// zones where Google Pub/Sub Lite is available.
+	Zone string
+}
+
+func (l LocationPath) String() string {
+	return fmt.Sprintf("projects/%s/locations/%s", l.Project, l.Zone)
+}
+
+// TopicPath stores the full path of a Google Pub/Sub Lite topic.
+// See https://cloud.google.com/pubsub/lite/docs/topics for more information.
+type TopicPath struct {
+	// A Google Cloud project. The project ID (e.g. "my-project") or the project
+	// number (e.g. 987654321) can be provided.
+	Project string
+
+	// A Google Cloud zone, for example "us-central1-a".
+	// See https://cloud.google.com/pubsub/lite/docs/locations for the list of
+	// zones where Google Pub/Sub Lite is available.
+	Zone string
+
+	// The ID of the Google Pub/Sub Lite topic, for example "my-topic-name".
+	TopicID string
+}
+
+func (t TopicPath) String() string {
+	return fmt.Sprintf("projects/%s/locations/%s/topics/%s", t.Project, t.Zone, t.TopicID)
+}
+
+// Location returns the path of the parent location.
+func (t TopicPath) Location() LocationPath {
+	return LocationPath{Project: t.Project, Zone: t.Zone}
+}
+
+var topicPathRE = regexp.MustCompile(`^projects/([^/]+)/locations/([^/]+)/topics/([^/]+)$`)
```

Comment:
If we are checking for structure, perhaps we should also check for validity of the topic name? Resource names are [documented for Pub/Sub](https://cloud.google.com/pubsub/docs/admin#resource_names) but not sure about Lite.

Reply:
I think we should forward this to the server for validity checking, just in case the criteria changes in the future. I double-checked the server impl and it uses the same util function as Cloud Pub/Sub.